article/_posts/2019-06-11-Batch_PG.md

```diff
@@ -0,0 +1,96 @@
+
+*Online vs offline learning*: Online learning refers to the policy interacting with the environment. Offline refers to a policy learning from past explorations. Online learning can be done off or on-policy, but offline learning is necessarily off-policy.
+
+*Seq2Seq models*: The goal of Seq2Seq models in natural language processing (NLP) is to produce an output sequence $$y = [a_0, a_1, ...a_T]$$ from an input sequence $$x$$, where $$a_i \in A$$ where $$A$$ is a vocabulary of words. Typically, an encoder will represent the input sequence as an euclidian vector and a decoder will convert the vector into an output sequence. In an RL setting, the state at time $$t$$ will be the input sequence $$x$$ and the actions $$y_{t-1} = [a_0, a_1, ..., a_{T-1}]$$
+
+# Methods
+
+The goal is to maximize the following (1):
+
+$$ J(\theta) = \sum_{s \in S} q(s)V^{\pi_\theta}(s) $$
+
+where $$q$$ is the distribution of states. The policy is updated by taking steps along the gradient of the objective, 
+
+$${\nabla}J(\theta) = {\nabla}E_{s~q}[\sum_{a \in A}\pi_\theta(a|s)Q^{\pi_\theta}(s,a)] = E_{s~q}[\sum_{a \in A} \nabla\pi_\theta(a|s)Q^{\pi_\theta}(s,a) + \pi_\theta(a|s)\nabla Q^\pi_\theta(s,a)]$$
+
+Because the second term in the above gradient is hardly tractable, it can be ignored and the gradient can be approximated to:
+
+$$g(\theta) = E_{s~q}[\sum_{a\in A}\nabla\pi_\theta(a|s)Q^{\pi_\theta}(s,a)]$$
+
+Expanding on $$g(\theta)$$, we get:
+
+![](/article/images/batchpg/2.jpeg)
+
+with 
+$$\psi(a,s)=\frac{\nabla_{\pi_{\theta}}(a|s)}{\pi_\theta(a|s)}$$ the score function of the policy 
+$$rho(s,a)=\frac{\pi_\theta(a|s)}{q(a|s)}$$a the importance sampling coefficient.
+
+The value function $$V^{\pi_\theta}$$ was introduced to reduce variance during the SGD updates: If $$Q^{\pi_\theta}(s_t,a_t)$$ is large relatively to $$V^{\pi_\theta}(s_t)$$, the coefficient of the score function will influence changes to $$\theta$$ to assign a higher probability to $$a_t$$ at $$s_t$$
+
+However, $$Q^{\pi_\theta}$$ and $$V^{\pi_\theta}$$ aren't available. $$Q^{\pi_\theta}(s_t,a_t)$$ can be approximated by summing the future rewards at the current state. However, since trajectories come from $$q$$ and not $$\pi_\theta$$, the authors use $$\lambda-return$$, a convex combination of the re-weighted rewards defined recursively from the end as
+
+![](/article/images/batchpg/3.jpeg)
+
+$$V^{\pi_\theta}$$ still needs to be defined. Approximating it can be quite difficult at $$S$$ can be potentially infinite. However, the role of $$V$$ is only to reduce variance and does not need to be exact or even precise. Two options are proposed to approximate $$V$$: The first one being a constant equal to the mean reward of the entire dataset. The second option uses the parametrisation $$V(s)=\sigma(\xi^T\phi(s))$$ estimated using the $$GTD(\lambda)$$ estimator. Their implementation of it is available in the appendix of the paper.
```

Comment:
Approximating it can be quite difficult at $$S$$ can be potentially infinite.: Approximating it can be quite difficult as $$S$$ can be potentially infinite.

Reply:
Fixed in ec6b13a.